File: elixir/lib/2022/04_camp_cleanup/aoc202204.ex
defmodule AOC2022.Day04 do
  @moduledoc """
  Advent of Code 2022, day 4: Camp Cleanup.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input),
    do: puzzle_input |> String.split("\n", trim: true) |> Enum.map(&parse_pair/1)

  @doc """
  Parse one pair of elf assignments.

  ## Example:

      iex> parse_pair("4-11,3-5")
      {4..11, 3..5}
  """
  def parse_pair(line) do
    line
    |> String.split(",")
    |> then(fn [first, second] -> {parse_elf(first), parse_elf(second)} end)
  end

  @doc """
  Parse one elf assignment.

  ## Example:

      iex> parse_elf("5-19")
      5..19
  """
  def parse_elf(pair) do
    pair
    |> String.split("-")
    |> then(fn [low, high] -> String.to_integer(low)..String.to_integer(high) end)
  end

  @doc """
  Solve part 1.
  """
  def part1(pairs), do: pairs |> Enum.count(&all_overlap?/1)

  @doc """
  Solve part 2.
  """
  def part2(pairs), do: pairs |> Enum.count(&any_overlap?/1)

  @doc """
  Check if a pair of assignments overlap in some way.

  ## Examples:

      iex> overlap?(5..19, 1..7, &Enum.all?/2)
      false
      iex> overlap?(5..19, 1..7, &Enum.any?/2)
      true
      iex> overlap?(7..7, 6..8, &Enum.all?/2)
      true
  """
  def overlap?(small, big, any_all), do: small |> any_all.(fn id -> id in big end)

  @doc """
  Check if one assignment in a pair completely overlaps the other.

  ## Examples:

      iex> all_overlap?({5..7, 5..7})
      true
      iex> all_overlap?({1..4, 5..9})
      false
      iex> all_overlap?({1..100, 42..42})
      true
  """
  def all_overlap?({first, second}),
    do: overlap?(first, second, &Enum.all?/2) || overlap?(second, first, &Enum.all?/2)

  @doc """
  Check if there are any overlaps in the assignments of a pair of elves.

  ## Examples:

      iex> any_overlap?({5..7, 5..7})
      true
      iex> any_overlap?({1..4, 5..9})
      false
      iex> any_overlap?({1..42, 42..100})
      true
  """
  def any_overlap?({first, second}),
    do: overlap?(first, second, &Enum.any?/2) || overlap?(second, first, &Enum.any?/2)

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2022/04_camp_cleanup/aoc202204.py
"""AoC 4, 2022: Camp Cleanup."""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return [parse_pair(line) for line in puzzle_input.split("\n")]


def parse_pair(line):
    """Parse one pair of assignments.

    ## Example:

    >>> parse_pair("5-7,6-9")
    ((5, 7), (6, 9))
    """
    return tuple(tuple(int(ids) for ids in elf.split("-")) for elf in line.split(","))


def part1(pairs):
    """Solve part 1."""
    return sum(full_overlap(first, second) for first, second in pairs)


def part2(pairs):
    """Solve part 2."""
    return sum(any_overlap(first, second) for first, second in pairs)


def full_overlap(first, second):
    """Does one assignment fully overlap the other?

    ## Examples:

    >>> full_overlap((5, 7), (6, 9))
    False
    >>> full_overlap((4, 11), (6, 9))
    True
    """
    low_1, high_1 = first
    low_2, high_2 = second
    return (low_1 <= low_2 and high_1 >= high_2) or (
        low_2 <= low_1 and high_2 >= high_1
    )


def any_overlap(first, second):
    """Do the assignments overlap at all?

    ## Examples:

    >>> any_overlap((5, 7), (6, 9))
    True
    >>> any_overlap((1, 3), (4, 7))
    False
    """
    low_1, high_1 = first
    low_2, high_2 = second
    return low_1 <= high_2 and low_2 <= high_1


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
